File: pdb/h2o_calc.pl
#!/usr/bin/perl

# ************************************************************
# * Script permettant de générer une molécule d'eau à partir *
# * de calculs lorsque l'on connait la taille des liaisons   *
# * et les angles                                            *
# * Auteur : Sébastien CELLES                                *
# * Licence : GNU General Public License v2 and above        *
# ************************************************************

# Ouverture du fichier en mode écriture (>) ou affichage d'erreur
open(FILE,">h2o_calc.pdb") or die("Erreur d'écriture");

# Définition de la fonction qui ajoute un atome dans le fichier .pdb
sub print_atom {
($AtomNumber,$AtomSymbol,$x,$y,$z)=($_[0],$_[1],$_[2],$_[3],$_[4]);
  printf(FILE "ATOM      %d  %s                  % 1.3f  % 1.3f  % 1.3f"."\n",
	 $AtomNumber,$AtomSymbol,$x,$y,$z);
# perldoc -f sprintf  pour %x.yf
# x longueur *totale* *minimale* du champ contenant un nombre
#  (et non celle de sa partie entière)
# y longueur partie décimale
}

# Calcul de pi (petite astuce vieille comme le monde informatique)  
$pi = 4*atan2(1, 1);

# Définition des constantes
#    Liaison O-H  0.0957nm = 0.957A (et pas 0,957)
#    Angle 104.52°
$l_OH = 0.957;
$alpha = 104.52*$pi/180; # angle en rad

# En tête du fichier .pdb
print FILE "HEADER"."\n";
print FILE "COMPND    H2O calc"."\n";
print FILE "HEAD"."\n";


# Partie centrale du fichier .pdb (ex : molécule d'eau)
#&print_atom(numéro_atome_dans_le_fichier_pdb,symbole_de_l_atome,x,y,z);
&print_atom(1,"O",0,0,0); # ne pas confondre O et 0
&print_atom(2,"H",$l_OH,0,0);
&print_atom(3,"H",$l_OH*cos($alpha),$l_OH*sin($alpha),0);

print FILE "CONECT    1    2    3"."\n";


# Fin du fichier .pdb
print FILE "END"."\n";
print FILE "TAIL"."\n";

close(FILE);

File: pdb/h2o_calc.py
#!/usr/bin/env python

# ************************************************************
# * Script en Python permettant de generer une molecule d'eau*
# * a partir de calculs lorsque l'on connait la taille des   *
# * liaisons et les angles                                   *
# * Auteur : Sebastien CELLES                                *
# * Licence : GNU General Public License v2 and above        *
# ************************************************************

from math import * # pi cos sin ...

FILE = open('h2o_calc.pdb', 'w')

# Definition de la fonction qui ajoute un atome dans le fichier .pdb
def print_atom(FILE, AtomNumber, AtomSymbol, x, y, z):
        str = 'ATOM      %(#1)s  %(#2)s                  %(#3) 1.3f  %(#4) 1.3f  %(#5) 1.3f' % {'#1': AtomNumber, '#2':AtomSymbol, '#3': x, '#4': y, '#5': z}
        FILE.write(str+"\n")

# Definition des constantes
#    Liaison O-H  0.0957nm = 0.957A (et pas 0,957)
#    Angle 104.52°
l_OH = 0.957;
alpha = 104.52*pi/180; # angle en rad

# En tete du fichier .pdb
FILE.write("HEADER"+"\n")
FILE.write("COMPND    H2O calc"+"\n")
FILE.write("HEAD"+"\n")

# Partie centrale du fichier .pdb (ex : molecule d'eau)
#print_atom(numéro_atome_dans_le_fichier_pdb,symbole_de_l_atome,x,y,z);
print_atom(FILE, 1, "O", 0, 0, 0) # ne pas confondre O et 0
print_atom(FILE, 2, "H", l_OH, 0, 0)
print_atom(FILE, 3, "H", l_OH*cos(alpha), l_OH*sin(alpha), 0)

#liaison(s)
FILE.write("CONECT    1    2    3"+"\n")

# Fin du fichier .pdb
FILE.write("END"+"\n")
FILE.write("TAIL"+"\n")

# Fermeture du fichier
FILE.close()
